### PR TITLE
Export tags through ?tag all --json

### LINKED ISSUE #83: Export tags
Hello!

Awhile back, members of The Coding Den asked for a way to export tags. This was implemented through `?tag all --text`, which uploads a parsable file containing the output of a `SELECT` query. However, I noticed that this doesn't include the content of the tags, meaning we can't export them directly to another bot, or to a JSON file. Is there a roadblock that's stopping this from happening?

### PR DESCRIPTION
**Pull request by OoLunar**
Same as `?tag all --text`, except it exports the content as JSON, with the tag's content. Closes #83. Example export:
```json
{"id":3,"name":"codeblocks","owner_id":336733686529654798,"uses":1,"can_delete":true,"is_alias":false,"content":"Use codeblocks for formatting code:\n\\`\\`\\`<language name>\n<your code here>\n\\`\\`\\`\n\nFor example:\n\\`\\`\\`python\nif True:\n    print(\"Hi!\")\n\\`\\`\\`\nProduces:\n```python\nif True:\n    print(\"Hi!\")\n```"}
```

### CODE REVIEW

File: cogs/tags.py [hunk 1]
```diff
@@ -870,6 +871,33 @@ async def _tag_all_text_mode(self, ctx):
...
+                """
+
+        bypass_owner_check = ctx.author.id == self.bot.owner_id or ctx.author.guild_permissions.manage_messages
+        rows = await ctx.db.fetch(query, ctx.guild.id, bypass_owner_check, ctx.author.id)
+        if not rows:
+            return await ctx.send('This server has no server-specific tags.')
+
+        for r in rows:
+            fp = io.BytesIO(r[0].encode())
+        await ctx.send(file=discord.File(fp, 'tags.txt'))
```
Comment (JeremyEspresso) [inline cogs/tags.py]: Shouldn't the extension be .json?

**Review by JeremyEspresso — CHANGES_REQUESTED** [review]

File: cogs/tags.py (hunk 1 above)
Comment (OoLunar) [reply cogs/tags.py]: Absolutely. I thought I did it, must've Ctrl + Z'd it

**Comment by OoLunar** [comment]
~~Just realized that `?tag all --json` only exports the latest tag instead of all of them. Fixing that now.~~

All tags are exported. Aliases are exported with the tags content.

**Comment by bsrecker** [comment]
I only see the latest tag being exported instead of all of the tags with this code change.

**Review by bsrecker — APPROVED** [review]
Downloaded and verified that changes work for tag export to json.

File: cogs/tags.py [hunk 2]
```diff
@@ -841,6 +841,7 @@ async def tags(self, ctx, *, member: TagMember = None):
     def _get_tag_all_arguments(args):
         parser = Arguments(add_help=False, allow_abbrev=False)
         parser.add_argument('--text', action='store_true')
+        parser.add_argument('--json', action='store_true')
```
Comment (AbstractUmbra) [inline cogs/tags.py]: This should perhaps be use [`ArgumentParser.add_mutually_exclusive_group`](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group) as currently you can use both flags and only the first will be accepted.

**Review by AbstractUmbra — CHANGES_REQUESTED** [review]

File: cogs/tags.py (hunk 2 above)
Comment (OoLunar) [reply cogs/tags.py]: Perhaps something like this? I don't typically program in Python, so I'm sure what I'm doing can be optimized. Thank you very much for linking the docs
```python
def _get_tag_all_arguments(args):
    parser = Arguments(add_help=False, allow_abbrev=False)
    parser.add_mutually_exclusive_group(required=False)
    parser.add_argument('--text', action='store_true')
    parser.add_argument('--json', action='store_true')
    if args is not None:
        return parser.parse_args(shlex.split(args))
    else:
        return parser.parse_args([])
 ```

Comment (AbstractUmbra) [reply cogs/tags.py]: `add_mutually_exclusive_group` returns the `group` that you should add the arguments to:

```py
    group = parser.add_mutually_exclusive_group(required=False)
    group.add_argument('--text', action='store_true')
    group.add_argument('--json', action='store_true')
```

Comment (OoLunar) [reply cogs/tags.py]: Just tested your suggestion. Everything seems good, I'll put it into the PR now

**Comment by thecashewtrader** [comment]
Any update on this PR? This is a feature that would greatly help me.

**Comment by OoLunar** [comment]
From what I understand, R. Danny is no longer being worked on and is only functioning until the message intent kills it. I don't think the PR will be merged

**Comment by thecashewtrader** [comment]
That's sad tbh, R. Danny being phased out was one of the main reasons I wanted an export, so that I can have tags backed up in plaintext.

**Comment by JDJGInc** [comment]
> From what I understand, R. Danny is no longer being worked on and is only functioning until the message intent kills it. I don't think the PR will be merged

R.danny has been verified, I think this pr is fine to be merged if Danny likes the concept and code.

**Pull request closed without merging**